python/ambassador/ir/irutils.py

```diff
@@ -181,14 +184,21 @@ def selector_matches(
         logger.debug("    no incoming labels => False")
         return False
 
-    selmatch = False
+    # Ambassador (2.0-2.3) & (3.0-3.1) consider a match on a single label as a "good enough" match.
+    # In versions 2.4+ and 3.2+ _ALL_ labels in a selector must be present for it to be considered a match.
+    # DISABLE_STRICT_LABEL_SELECTORS provides a way to restore the old unintended loose matching behaviour
+    # in the event that it is desired. The ability to disable strict label matching will be removed in a future version.
+    disable_strict_selectors = parse_bool(os.environ.get("DISABLE_STRICT_LABEL_SELECTORS", "false"))
 
+    # For every label in mappingSelector, there must be a label with same value in Mapping itself.
     for k, v in match.items():
         if labels.get(k) == v:
             logger.debug("    selector match for %s=%s => True", k, v)
-            return True
-
-        logger.debug("    selector miss on %s=%s", k, v)
+            if disable_strict_selectors:
+                return True
+        elif not disable_strict_selectors:
+            logger.debug("    selector miss for %s=%s => False", k, v)
+            return False
```

Comment:
Maybe it's just me, but I had to twist my brain around to understand this.  Perhaps don't try so hard to share the code?

```python
if disable_strict_selectors:
    old code goes here
else:
    new code goes here
```